src/EventIndexing.js

```diff
@@ -0,0 +1,406 @@
+            return;
+        }
+
+        const e = ev.toJSON().decrypted;
+        const profile = {
+            displayname: ev.sender.rawDisplayName,
+            avatar_url: ev.sender.getMxcAvatarUrl(),
+        };
+
+        indexManager.addEventToIndex(e, profile);
+    }
+
+    async crawlerFunc(handle) {
+        // TODO either put this in a better place or find a library provided
+        // method that does this.
+        const sleep = async (ms) => {
+            return new Promise(resolve => setTimeout(resolve, ms));
+        };
+
+        let cancelled = false;
+
+        console.log("EventIndex: Started crawler function");
+
+        const client = MatrixClientPeg.get();
+        const indexManager = PlatformPeg.get().getEventIndexingManager();
+
+        handle.cancel = () => {
+            cancelled = true;
+        };
+
+        while (!cancelled) {
+            // This is a low priority task and we don't want to spam our
+            // Homeserver with /messages requests so we set a hefty timeout
+            // here.
+            await sleep(this._crawlerTimeout);
+
+            console.log("EventIndex: Running the crawler loop.");
+
+            if (cancelled) {
+                break;
+            }
+
+            const checkpoint = this.crawlerCheckpoints.shift();
+
+            /// There is no checkpoint available currently, one may appear if
+            // a sync with limited room timelines happens, so go back to sleep.
+            if (checkpoint === undefined) {
+                continue;
+            }
+
+            console.log("EventIndex: crawling using checkpoint", checkpoint);
+
+            // We have a checkpoint, let us fetch some messages, again, very
+            // conservatively to not bother our Homeserver too much.
+            const eventMapper = client.getEventMapper();
+            // TODO we need to ensure to use member lazy loading with this
+            // request so we get the correct profiles.
+            let res;
+
+            try {
+                res = await client._createMessagesRequest(
+                    checkpoint.roomId, checkpoint.token, this._eventsPerCrawl,
+                    checkpoint.direction);
+            } catch (e) {
+                console.log("EventIndex: Error crawling events:", e);
+                this.crawlerCheckpoints.push(checkpoint);
+                continue;
+            }
+
+            if (res.chunk.length === 0) {
+                console.log("EventIndex: Done with the checkpoint", checkpoint);
+                // We got to the start/end of our timeline, lets just
+                // delete our checkpoint and go back to sleep.
+                await indexManager.removeCrawlerCheckpoint(checkpoint);
+                continue;
+            }
+
+            // Convert the plain JSON events into Matrix events so they get
+            // decrypted if necessary.
+            const matrixEvents = res.chunk.map(eventMapper);
```

Comment:
I do feel like you're rebuilding a lot of the JS SDK all over again... I don't have a concrete suggestion for the moment, but something to keep in mind.